parsers/markdown.c

```diff
@@ -259,6 +268,94 @@ static void notifyEndOfCodeBlock (markdownSubparser *m)
 	}
 }
 
+typedef enum {
+	HTAG_SPACE_FOUND,
+	HTAG_HASHTAG_FOUND,
+	HTAG_TEXT,
+} hashtagState;
+
+/*
+   State machine to find all hashtags in a line.
+ */
+static void getAllHashTagsInLineMaybe (const unsigned char *line, int pos,
+	int lineLen, hashtagState state)
+{
+	while (pos < lineLen)
+	{
+		switch (state)
+		{
+		case HTAG_SPACE_FOUND:
+			if (line[pos] == '#')
+			{
+				state = HTAG_HASHTAG_FOUND;
+			}
+			else if (!isspace (line[pos]))
+				state = HTAG_TEXT;
+			pos++;
+			break;
+		case HTAG_HASHTAG_FOUND:
+		{
+			/* #123 is invalid */
+			bool hasNonNumericalChar = false;
+			int start = pos;
+			while (pos < lineLen)
+			{
+				int utf8_len;
+				if (isalpha (line[pos])
+					|| line[pos] == '_' || line[pos] == '-' || line[pos] == '/')
+				{
+					hasNonNumericalChar = true;
+					pos++;
+				}
+				else if (isdigit (line[pos]))
+				{
+					pos++;
+				}
+				else if ((utf8_len =
+						utf8_raw_strlen ((const char *) &line[pos],
+							lineLen - pos)) > 0)
+				{
+					hasNonNumericalChar = true;
+					pos += utf8_len;
+					Assert (pos <= lineLen);
+				}
+				else
+				{
+					break;
+				}
+			}
+
+			int length = pos - start;
+			if (hasNonNumericalChar && length > 0)
+			{
+				vString *tag =
+					vStringNewNInit ((const char *) (&(line[start])), length);
+				int r = makeMarkdownTag (tag, K_HASHTAG, false);
+				attachParserFieldToCorkEntry (r,
+					MarkdownFields[F_HASHTAG].ftype, "#");
+				nestingLevelsPush (nestingLevels, r);
```

Comment:
(review comment Group 2)

Pushed but never poped :-)


The `nestingLevels` is a variable for filling the scope field:

```
$ cat /tmp/input.md 
# section 1

## sub section 2

## sub section 3

# section A

## sub section B

## sub section C
$ ctags -o - --sort=no --fields-Markdown='{sectionMarker}' input.md
section 1	input.md	/^# section 1$/;"	c	sectionMarker:#
sub section 2	input.md	/^## sub section 2$/;"	s	chapter:section 1	sectionMarker:#
sub section 3	input.md	/^## sub section 3$/;"	s	chapter:section 1	sectionMarker:#
section A	input.md	/^# section A$/;"	c	sectionMarker:#
sub section B	input.md	/^## sub section B$/;"	s	chapter:section A	sectionMarker:#
sub section C	input.md	/^## sub section C$/;"	s	chapter:section A	sectionMarker:#
```

In my understanding, a hashtag doesn't make a scope.
So you don't have to push the tag (a value returned from `makeMarkdownTag`) for a hashtag to the `nestingLevels`. 

Your change makes one tag `a/b` for the input "#a/b".
If I make a pull request for the same purpose,
for aligning this pull request to the design of Universal Ctags, I will make two tags (`a` and `b`), and one full-qualified extra tag (`a/b`) for the input. `b` may have `hashtag:a` as the value for its scope field:
```
a	input.md	/^... #a/b ...$/;"	h
b	input.md	/^... #a/b ...$/;"	h	hashtag:a
```

This may help a client tool show a "tree view" of tags.

The full-qualified extra tag is emitted only when --extras=+q option is given.
With the option, in addition to the above `a` and `b`, ctags with my imaginary pull request, will emit:
```
a/b	input.md	/^... #a/b ...$/;"	h	extras:qualified
```

The `nestingLevels` is useful to tell ctags the relation between `a` and `b`; making `b` in the environment where `a` is pushed to the `nestingLevels`, ctags fills the scope field of `b` with `a` automatically. In addition, ctags also makes `a/b`.

Anyway, the current pull request is enough useful; yes, you are the user!
If you are interested in the fine-grained tagging, we can try implementing it after merging this:)

Reply:
Thank you for your meticulousness and patience, I previously didn't understand the details of the code thoroughly enough. I will  attempt to implement it later on.